📖 Define testing guidelines and standards (#3349)
kubernetes-sigs/cluster-api: docs/book/src/developer/testing.md

[envtest] is a testing environment that is provided by the [controller-runtime] project. This environment spins up a
local instance of etcd and the kube-apiserver. This allows test to be executed in an environment very similar to a
real environment. 

Additionally, in Cluster API there is a set of utilities under [test/helpers] that helps developers in setting up
a [envtest] ready for Cluster API testing, and most specifically:

- With the required CRDs already pre-configured.
- With all the Cluster API webhook pre-configured, so there are enforced guarantees about the semantic accuracy
  of the test objects you are going to create.
  
Most notably, [envtest] does not provides only a real API server to user during test, but it offers the opportunity
to configure one or more controllers to run against the test cluster; by using this feature it is possible to use
[envtest] for developing Cluster API integration tests.

Please note that, because [envtest] uses a real kube-apiserver that is shared across many tests, the developer
should take care of ensuring each test run in isolation from the others, by:
 
- Creating objects in separated namespaces.
- Avoiding object name conflict.

> - Cleaning up objects created by each test.

Reviewer: I'm guessing initially we will have an area of the codebase with tests written in the preferred style. This will probably result in creation of additional helpers that will help manage this overhead. These helpers should then be used everywhere else in order to provide a consistent test codebase.

Author: This makes totally sense

agreement among Cluster API maintainers that usage [fakeclient] should be progressively deprecated in favor of usage
of [envtest].

### `ginkgo`
[Ginkgo] is a Go testing framework built to help you efficiently write expressive and comprehensive tests using Behavior-Driven Development (“BDD”) style.

While [Ginkgo] is widely used in the Kubernetes ecosystem, Cluster API maintainers found the lack of integration with the 
most used golang IDE somehow limiting, mostly because:

- it makes interactive debugging of tests more difficult, since you can't just run the test using the debugger directly
- it makes it more difficult to only run a subset of tests, since you can't just run or debug individual tests using an IDE, 
  but you now need to run the tests using `make` or the `ginkgo` command line and override the focus to select individual tests
  
In Cluster API you MUST use ginkgo only for E2E tests, where it is required to leverage on the support for running specs
in parallel; in any case, developers MUST NOT use the table driven extension DSL (`DescribeTable`, `Entry` commands)

> which is consider un-intuitive.

Reviewer: nit: 
```suggestion
which is considered unintuitive.
```

- With all the Cluster API webhook pre-configured, so there are enforced guarantees about the semantic accuracy
  of the test objects you are going to create.
  

> This is an example of how to create an instance of [envtest] that can be shared across all the tests in a package:

Author: I don't have strong preferences.. @vincepri opinions?

> This is an example of how to create an instance of [envtest] that can be shared across all the tests in a package:

Reviewer 2: `suite_test.go` sounds good to me

 
```golang
var (
	testEnv *helpers.TestEnvironment
	ctx     = context.Background()
)

func TestMain(m *testing.M) {
	// Bootstrapping test environment
	testEnv = helpers.NewTestEnvironment()
	go func() {
		if err := testEnv.StartManager(); err != nil {
			panic(fmt.Sprintf("Failed to start the envtest manager: %v", err))
		}
	}()
	// Run tests
	code := m.Run()
	// Tearing down the test environment
	if err := testEnv.Stop(); err != nil {
		panic(fmt.Sprintf("Failed to stop the envtest: %v", err))
	}

	// Report exit code
	os.Exit(code)
}
```
  

> Most notably, [envtest] does not provides only a real API server to user during test, but it offers the opportunity

Reviewer 3: ```suggestion
Most notably, [envtest] provides not only a real API server to user during test, but it offers the opportunity
```

- Avoiding object name conflict.
- Cleaning up objects created by each test.

> This is an example of a test implementing those recommendations:
> 
> ```golang
> func TestAFunc(t *testing.T) {
> 	g := NewWithT(t)
> 	// Generate namespace with a random name starting with ns1; such namespace
> 	// will host test objects in isolation from other tests.
> 	ns1, err := testEnv.CreateNamespace(ctx, "ns1") 
> 	g.Expect(err).ToNot(HaveOccurred())
> 	defer func() {
> 		// Cleanup the test namespace
> 		g.Expect(testEnv.DeleteNamespace(ctx, ns1)).To(Succeed())
> 	}()
> 
> 	obj := &clusterv1.Cluster{
> 		ObjectMeta: metav1.ObjectMeta{
> 			Name:      "test",
> 			Namespace: ns1.Name, // Place test objects in the test namespace
> 		},
> 	}
> 	
> 	// Actual test code...
> }
> ```
> 
> In case of object used in many test case withing the same test, it is possible to leverage on Kuberentes `GenerateName`:
> 
> ```golang
> func TestAFunc(t *testing.T) {
> 	// create test namespace
> 
> 	obj := &clusterv1.Cluster{
> 		ObjectMeta: metav1.ObjectMeta{
> 			GenerateName: "test-",  // Instead of assigning a name, use GenerateName
> 			Namespace:    ns1.Name, // Place test objects in the test namespace
> 		},
> 	}
> 	
> 	t.Run("test case 1", func(t *testing.T) {
> 		g := NewWithT(t)
> 		// Deep copy the object in each test case, so we prevent side effects in case the object changes.
> 		// Additionally, thanks to GenerateName, the objects gets a new name for each test case.
> 		obj := obj.DeepCopy()
> 
> 	    // Actual test case code...
> 	}
> 	t.Run("test case 2", func(t *testing.T) {
> 		g := NewWithT(t)
> 		obj := obj.DeepCopy()
> 
> 	    // Actual test case code...
> 	}
> 	// More test cases.
> }
> ```

Reviewer: I think we can combine the two examples into one.
```suggestion
This is an example of a test implementing those recommendations.
We can leverage the `GenerateName` property to generate unique names for our objects.
For objects that are shared across sub-tests, ensure they are scoped within the test namespace and deep copied to avoid cross-test changes that may occur to the object.

```golang
func TestAFunc(t *testing.T) {
	g := NewWithT(t)
	// Generate namespace with a random name starting with ns1; such namespace
	// will host test objects in isolation from other tests.
	ns1, err := testEnv.CreateNamespace(ctx, "ns1") 
	g.Expect(err).ToNot(HaveOccurred())
	defer func() {
		// Cleanup the test namespace
		g.Expect(testEnv.DeleteNamespace(ctx, ns1)).To(Succeed())
	}()

	obj := &clusterv1.Cluster{
		ObjectMeta: metav1.ObjectMeta{
			GenerateName: "test-",  // Instead of assigning a name, use GenerateName
			Namespace: ns1.Name, // Place test objects in the test namespace
		},
	}
		
	t.Run("test case 1", func(t *testing.T) {
		g := NewWithT(t)
		// Deep copy the object in each test case, so we prevent side effects in case the object changes.
		// Additionally, thanks to GenerateName, the objects gets a new name for each test case.
		obj := obj.DeepCopy()

	        // Actual test case code...
	}
	t.Run("test case 2", func(t *testing.T) {
		g := NewWithT(t)
		obj := obj.DeepCopy()

	        // Actual test case code...
	}
	// More test cases.
}
```
```

Author: I prefer to keep the two examples separated so we have both an example for simple test and the example for table-driven tests. Never the less I took your sentence about deep copying objects and included in the doc
